File: edit_distance.rb
# RUBY

module EditDistance
  # Naive recursive levenshtein edit distance.
  def levenshtein(a, b)
    return b.size if a.size == 0
    return a.size if b.size == 0
    [(levenshtein a, b.chop) + 1,
      (levenshtein a.chop, b) + 1,
      (levenshtein a.chop, b.chop) + (a[-1] == b[-1] ? 0 : 1)].min
  end

  # Dynamic programming top-down levenshtein edit distance.
  def memoized_levenshtein(a, b)
    d = []
    (a.size + 1).times do |i|
      d[i] = []
    end
    lookup_levenshtein a, b, d
  end

  private

  def lookup_levenshtein(a, b, d)
    i, j = a.size, b.size
    return d[i][j] unless d[i][j].nil?
    if i == 0 || j == 0
      d[i][j] = [i, j].max
    else
      actions = [(lookup_levenshtein a, b.chop, d) + 1,
                (lookup_levenshtein a.chop, b, d) + 1,
                (lookup_levenshtein a.chop, b.chop, d) + (a[-1] == b[-1] ? 0 : 1)]
      d[i][j] = actions.min
    end
    d[i][j]
  end
end

File: edit_distance.py
# Naive recursive implementation.
def levenshtein(a, b):
	if len(a) == 0:
		return len(b)
	if len(b) == 0:
		return len(a)
	return min([levenshtein(a, b[:-1]) + 1,
			levenshtein(a[:-1], b) + 1,
			levenshtein(a[:-1], b[:-1]) + (a[-1] != b[-1])])

# Memoized recursive implementation.
def memoized_levenshtein(a, b):
	i, j = len(a), len(b)
	d = [[-1 for x in range(j + 1)] for y in range(i + 1)]
	return lookup_levenshtein(d, a, b)

def lookup_levenshtein(d, a, b):
	i, j = len(a), len(b)
	if d[i][j] != -1:
		return d[i][j]
	if i == 0:
		d[i][j] = j
	elif j == 0:
		d[i][j] = i
	else:
		substitution = a[-1] != b[-1]
		insert = lookup_levenshtein(d, a, b[:-1]) + 1
		delete = lookup_levenshtein(d, a[:-1], b) + 1
		substitute = lookup_levenshtein(d, a[:-1], b[:-1]) + substitution
		d[i][j] = min([insert, delete, substitute])
	return d[i][j]

a = "abcddd"
b = "abdce"
print a + ", " + b + ": Cost is " + str(memoized_levenshtein(a, b)) + "."